### PR TITLE
Purge all data for already deleted project #454

### PR DESCRIPTION
**Pull request by logiclord**
Azkaban already has method cleanOlderProjectVersion in place to hard delete a project version (https://github.com/azkaban/azkaban/blob/master/azkaban-common/src/main/java/azkaban/project/JdbcProjectLoader.java#L1038), which is used by Azkaban to do house keeping when a new project version is uploaded (https://github.com/azkaban/azkaban/blob/master/azkaban-common/src/main/java/azkaban/project/ProjectManager.java#L496). I have leveraged this exisitng functionality.

### CODE REVIEW

File: azkaban-common/src/main/java/azkaban/project/ProjectManager.java [hunk 1]
```diff
@@ -204,6 +204,34 @@ public Project getProject(int id) {
     return projectsById.get(id);
   }
 
+    /**
+     * fetch inactive project from db by project_id
+     *
+     * @param name
+     * @return
+     */
+    public Project getInactiveProject(String name) {
```
Comment (evlstyle) [inline azkaban-common/src/main/java/azkaban/project/ProjectManager.java]: function name is misleading, actually the function here will get a project with name match disregard the status, so it is not necessary to be Inactive.  Recommend to rename this to GetProject or GetProjectByName, state it in the meta that the returning result disregards the project status.  

Same comment for the function below, which gets the project by Id.

Comment (evlstyle) [inline azkaban-common/src/main/java/azkaban/project/ProjectManager.java]: function name is misleading, actually the function here will fetch a project with a successful name match, no matter it is active or not .  Simple fix here is to check the project and make sure it is inactive before returning, therefore the function works as it's name indicates. 
Same comment for the function below, which gets the project by Id.

File: azkaban-common/src/main/java/azkaban/project/JdbcProjectLoader.java [hunk 2]
```diff
@@ -169,6 +169,67 @@ private Project fetchProjectById(Connection connection, int id)
...
+        try {
+            project = fetchProjectByName(connection, name);
+        } finally {
+            DbUtils.closeQuietly(connection);
+        }
+
+        return project;
+    }
+
+    private Project fetchProjectByName(Connection connection, String name)
```
Comment (hluu) [inline azkaban-common/src/main/java/azkaban/project/JdbcProjectLoader.java]: I am surprised there isn't a method that does this already.

File: azkaban-common/src/main/java/azkaban/project/ProjectManager.java [hunk 3]
```diff
@@ -249,6 +277,25 @@ public Project createProject(String projectName, String description,
...
+     * of a project and log event in project_events table
+     *
+     * @param project
+     * @param deleter
+     * @return
+     * @throws ProjectManagerException
+     */
+    public synchronized Project purgeProject(Project project, User deleter)
+        throws ProjectManagerException {
+        projectLoader.cleanOlderProjectVersion(project.getId(),
```
Comment (hluu) [inline azkaban-common/src/main/java/azkaban/project/ProjectManager.java]: Should we check for null project and deleter parameters?

File: azkaban-common/src/test/java/azkaban/project/JdbcProjectLoaderTest.java [hunk 4]
```diff
@@ -214,6 +213,96 @@ private static void clearDB() {
     DbUtils.closeQuietly(connection);
   }
 
+    /** Test case to validated permissions for fetchProjectByName **/
+    @Test
+    public void testPermissionRetrivalByFetchProjectByName()
```
Comment (hluu) [inline azkaban-common/src/test/java/azkaban/project/JdbcProjectLoaderTest.java]: Great to see tests

Comment (hluu) [inline azkaban-common/src/test/java/azkaban/project/JdbcProjectLoaderTest.java]: seems like the method can be more clear in terms of what it is testing.

File: azkaban-common/src/test/java/azkaban/project/JdbcProjectLoaderTest.java [hunk 5]
```diff
@@ -214,6 +213,96 @@ private static void clearDB() {
...
+
+        Project fetchedProject = loader.fetchProjectByName(project.getName());
+        assertProjectMemberEquals(project, fetchedProject);
+        Assert.assertEquals(permOverride,
+            fetchedProject.getUserPermission(user.getUserId()));
+    }
+
+    /** Default Test case for fetchProjectByName **/
+    @Test
+    public void testProjectRetrivalByFetchProjectByName()
```
Comment (hluu) [inline azkaban-common/src/test/java/azkaban/project/JdbcProjectLoaderTest.java]: Retrival?  Do you mean Retrieval?

File: azkaban-common/src/test/java/azkaban/project/JdbcProjectLoaderTest.java [hunk 6]
```diff
@@ -214,6 +213,96 @@ private static void clearDB() {
...
+        if (!isTestSetup()) {
+            return;
+        }
+        ProjectLoader loader = createLoader();
+        try {
+            loader.fetchProjectByName("NonExistantProject");
+        } catch (ProjectManagerException ex) {
+            System.out.println("Test true");
+        }
+        Assert.fail("No exception");
```
Comment (hluu) [inline azkaban-common/src/test/java/azkaban/project/JdbcProjectLoaderTest.java]: Maybe make this message more clear i.e "Expecting exception, but didn't get one"

File: azkaban-webserver/src/main/java/azkaban/webapp/servlet/ProjectManagerServlet.java [hunk 7]
```diff
@@ -518,6 +520,50 @@ private void handleDownloadProject(HttpServletRequest req,
...
+
+        // project is already deleted
+        if (valid) {
+            project = projectManager.getInactiveProject(projectName);
+            // only eligible users can purge a project
+            if (!hasPermission(project, user, Type.ADMIN)) {
+                ret.put("error", "Cannot purge. User '" + user.getUserId()
+                    + "' is not an ADMIN.");
+                valid = false;
+                ;
```
Comment (hluu) [inline azkaban-webserver/src/main/java/azkaban/webapp/servlet/ProjectManagerServlet.java]: extra semicolon

File: azkaban-webserver/src/main/java/azkaban/webapp/servlet/ProjectManagerServlet.java [hunk 8]
```diff
@@ -518,6 +520,50 @@ private void handleDownloadProject(HttpServletRequest req,
...
+        if (valid) {
+            try {
+                projectManager.purgeProject(project, user);
+            } catch (ProjectManagerException e) {
+                ret.put("error", e.getMessage());
+                valid = false;
+            }
+        }
+
+        ret.put("success", valid);
```
Comment (hluu) [inline azkaban-webserver/src/main/java/azkaban/webapp/servlet/ProjectManagerServlet.java]: So the return object can have "success" as key and valid value as false.  Is that OK?

File: azkaban-common/src/main/java/azkaban/project/ProjectManager.java (hunk 1 above)
Comment (logiclord) [inline azkaban-common/src/main/java/azkaban/project/ProjectManager.java]: Yes, I named it InActive as there was already a method getProject to fetch active projects from Cache and to minimize code refractoring.
I have merged two methods and added another method to test if a project is Active.

File: azkaban-common/src/main/java/azkaban/project/ProjectManager.java (hunk 3 above)
Comment (logiclord) [inline azkaban-common/src/main/java/azkaban/project/ProjectManager.java]: I expect to throw exception in that scenario. For purge API, I have handled that case while invoking this method.

File: azkaban-common/src/test/java/azkaban/project/JdbcProjectLoaderTest.java (hunk 4 above)
Comment (logiclord) [inline azkaban-common/src/test/java/azkaban/project/JdbcProjectLoaderTest.java]: fixed

File: azkaban-common/src/test/java/azkaban/project/JdbcProjectLoaderTest.java (hunk 5 above)
Comment (logiclord) [inline azkaban-common/src/test/java/azkaban/project/JdbcProjectLoaderTest.java]: fixed

File: azkaban-common/src/test/java/azkaban/project/JdbcProjectLoaderTest.java (hunk 6 above)
Comment (logiclord) [inline azkaban-common/src/test/java/azkaban/project/JdbcProjectLoaderTest.java]: changed

File: azkaban-webserver/src/main/java/azkaban/webapp/servlet/ProjectManagerServlet.java (hunk 7 above)
Comment (logiclord) [inline azkaban-webserver/src/main/java/azkaban/webapp/servlet/ProjectManagerServlet.java]: removed

File: azkaban-webserver/src/main/java/azkaban/webapp/servlet/ProjectManagerServlet.java (hunk 8 above)
Comment (logiclord) [inline azkaban-webserver/src/main/java/azkaban/webapp/servlet/ProjectManagerServlet.java]: yes

**Comment by logiclord** [comment]
Incorporated feedback comments and updated the review.

File: azkaban-common/src/test/java/azkaban/project/JdbcProjectLoaderTest.java (hunk 4 above)
Comment (johnyu0520) [inline azkaban-common/src/test/java/azkaban/project/JdbcProjectLoaderTest.java]: nit picking, retrieval is mis-spelt :)

File: azkaban-common/src/test/java/azkaban/project/JdbcProjectLoaderTest.java [hunk 9]
```diff
@@ -214,6 +213,96 @@ private static void clearDB() {
     DbUtils.closeQuietly(connection);
   }
 
+    /** Test case to validated permissions for fetchProjectByName **/
+    @Test
+    public void testPermissionRetrivalByFetchProjectByName()
+        throws ProjectManagerException {
+        if (!isTestSetup()) {
+            return;
```
Comment (johnyu0520) [inline azkaban-common/src/test/java/azkaban/project/JdbcProjectLoaderTest.java]: just curious, if the test is not setup, should we setting up the test, or should we be returning success with the test?

File: azkaban-webserver/src/main/java/azkaban/webapp/servlet/ProjectManagerServlet.java [hunk 10]
```diff
@@ -518,6 +520,51 @@ private void handleDownloadProject(HttpServletRequest req,
...
+     * validate readiness of a project and user permission and use
+     * projectManager to purge the project if things looks good
+     **/
+    private void handlePurgeProject(HttpServletRequest req,
+        HttpServletResponse resp, Session session) throws ServletException,
+        IOException {
+        User user = session.getUser();
+        String projectName = getParam(req, "project");
+        HashMap<String, Object> ret = new HashMap<String, Object>();
+        boolean valid = true;
```
Comment (johnyu0520) [inline azkaban-webserver/src/main/java/azkaban/webapp/servlet/ProjectManagerServlet.java]: Just thinking, should the variable "valid" be renamed to something better?
Does it stand for "shouldPurge", or does it stand for "purgeOperationSuccessful"?

File: azkaban-webserver/src/main/java/azkaban/webapp/servlet/ProjectManagerServlet.java [hunk 11]
```diff
@@ -518,6 +520,51 @@ private void handleDownloadProject(HttpServletRequest req,
...
+
+        try {
+            Project project = null;
+
+            // project is already deleted
+            if (valid) {
+                project = projectManager.getProject(projectName);
+                // only eligible users can purge a project
+                if (project == null
+                    || !hasPermission(project, user, Type.ADMIN)) {
```
Comment (johnyu0520) [inline azkaban-webserver/src/main/java/azkaban/webapp/servlet/ProjectManagerServlet.java]: Looks like this if statement contains 2 possible scenarios.  One is project already deleted.  The other is user does not have admin permission.  Should they be sending out different error messages?

File: azkaban-common/src/test/java/azkaban/project/JdbcProjectLoaderTest.java (hunk 9 above)
Comment (logiclord) [inline azkaban-common/src/test/java/azkaban/project/JdbcProjectLoaderTest.java]: makes sense. incorporated the feedback

File: azkaban-webserver/src/main/java/azkaban/webapp/servlet/ProjectManagerServlet.java (hunk 10 above)
Comment (logiclord) [inline azkaban-webserver/src/main/java/azkaban/webapp/servlet/ProjectManagerServlet.java]: updated

File: azkaban-webserver/src/main/java/azkaban/webapp/servlet/ProjectManagerServlet.java (hunk 11 above)
Comment (logiclord) [inline azkaban-webserver/src/main/java/azkaban/webapp/servlet/ProjectManagerServlet.java]: updated

**Comment by logiclord** [comment]
incorporated feedback

**Comment by johnyu0520** [comment]
LGTM

**Pull request merged**